File: G00232918_James_Connolly_Assignment_1/python_proc/shop.py
from dataclasses import dataclass, field
from typing import List
import csv

# setting the data classes for shop elements
@dataclass
class Product:
    name: str
    price: float = 0.0

@dataclass 
class ProductStock:
    product: Product
    quantity: int

@dataclass 
class Shop:
    cash: float = 0.0
    stock: List[ProductStock] = field(default_factory=list)

@dataclass
class Customer:
    name: str = ""
    budget: float = 0.0
    shopping_list: List[ProductStock] = field(default_factory=list)


# function to create and stock the shop from the stock csv
def create_and_stock_shop():
    s = Shop()
    with open('G00232918_James_Connolly_Assignment_1\stock.csv') as csv_file:
        csv_reader = csv.reader(csv_file, delimiter=',')
        first_row = next(csv_reader)
        s.cash = float(first_row[0])
        for row in csv_reader:
            p = Product(row[0], float(row[1]))
            ps = ProductStock(p, float(row[2]))
            s.stock.append(ps)
    return s

# print the product info, name and quantity
def print_product(p):
    print(f'\nPRODUCT NAME: {p.name} \nPRODUCT PRICE: {p.price}')

# print the shop cash and quantities of stock
def print_shop(s):
    print(f'Shop has {s.cash} in cash')
    for item in s.stock:
        print_product(item.product)
        print(f'The Shop has {item.quantity} of the above')

# check shop stock
def check_prod_stock(n, quantity, shop):
    for item in shop.stock:
        if n == item.product.name:
            if quantity <= item.quantity:
                item.quantity -= quantity
                return quantity
            elif quantity > item.quantity:
                # adjust the quantity available
                quantity = int(item.quantity)
                item.quantity = 0
                # error message if not in stock
                raise ValueError(f"We don't have that many {n} in stock. We can only give you {quantity}")
    # error if the product cannot be found
    raise ValueError(f"We don't have {n} in stock.")

# reads in the customer lists created
def read_customer(file_path):
    with open(file_path) as csv_file:
        csv_reader = csv.reader(csv_file, delimiter=',')
        first_row = next(csv_reader)
        c = Customer(first_row[0], float(first_row[1]))
        for row in csv_reader:
            name = row[0]
            quantity = float(row[1])
            p = Product(name)
            ps = ProductStock(p, quantity)
            c.shopping_list.append(ps)
        return c

# function to to print customer information
def print_customer(customer, shop):
    print(f"{customer.name}'s shopping list")
    print(f"BUDGET: €{customer.budget:.2f}")
    total_bill = 0

    for item in customer.shopping_list:
        found_match = False
        for prod in shop.stock:
            # remove any unwanted white spaces 
            if item.product.name.strip().lower() == prod.product.name.strip().lower():
                # if the match is found
                found_match = True
                # record the name of match
                choice_name = item.product.name
                # record the price
                shop_price = prod.product.price
                # convert the quantity to an integer
                item.quantity = int(item.quantity)

                try:
                    quantity_purchased = check_prod_stock(prod.product.name, item.quantity, shop)
                    cost = quantity_purchased * shop_price 
                    prod.quantity -= quantity_purchased
                    # if the customer budget is less than the cost
                    if customer.budget < cost:
                        print(f"** Sorry! You don't have enough money left for {choice_name}! **")
                        continue
                    else:
                        # updating each cost        
                        total_bill += cost
                        customer.budget -= cost
                        shop.cash += cost

                    print(f"PRODUCT NAME: {item.product.name}\n")
                    print(f"PRODUCT PRICE: €{prod.product.price:.2f}\n")
                    print(f"QUANTITY REQUIRED: {item.quantity}\n")
                    print(f"QUANTITY PURCHASED: {quantity_purchased}\n")
                    print(f"TOTAL ITEM COST: €{cost:.2f}\n")
                    print(f"ADJUSTED BUDGET: €{customer.budget:.2f}\n")
                    print(f"(ADJUSTED SHOP FLOAT: €{shop.cash:.2f})\n")
                    
                # I put this in because I was getting duplicates when selecting customer csvs
                except ValueError as e:
                    print(f"** Warning: {e} **")
        # if no match            
        if not found_match:
            print(f"** Warning: Product {item.product.name} not found in the shop's stock! **")

    print(f"\nTOTAL BILL: €{total_bill:.2f}")
    print(f"BUDGET REMAINING: €{customer.budget:.2f}")
    print("\n** Thank you for your custom **\n")

    return customer


def liveMode(shop):
    budget = float(input("What is your budget? \n"))
    totalBill = 0
    print("Welcome to Procedural Python Shopping Program\n")
    while True:
        print(f"Your current budget is €{budget:.2f}.\n\nPlease select what you would like to buy from the list below:\n")
        for index, prod in enumerate(shop.stock, start=1):
            print(f"{index} - {prod.product.name} @ €{prod.product.price:.2f} each")

        print("*9* - Finish shopping and print total bill")
        print("*10* - Exit Live Mode\n")

        choice = int(input("Please make your selection: "))

        if choice == 9:
            print("\n--------------------\n")
            print("Come again soon and have a nice day!\n")
            print("--------------------\n")
            break

        elif 1 <= choice <= len(shop.stock):
            choice -= 1
            choice_name = shop.stock[choice].product.name
            choice_price = shop.stock[choice].product.price
            choice_detail = shop.stock[choice].product
            quant = int(input(f"How many {choice_name} would you like to purchase? "))
            purch_quan = check_prod_stock(choice_name, quant, shop)
            totalCost = purch_quan * choice_price

            if budget < totalCost:
                print(f"\n** You don't have enough money for {choice_name}! **\n")
                continue

            budget -= totalCost
            totalBill += totalCost
            Shop.cash += totalBill

            print_product(choice_detail)

            print(f"QUANTITY REQUIRED: {quant}")
            print(f"QUANTITY PURCHASED: {purch_quan}")
            print(f"TOTAL ITEM COST: €{totalCost:.2f}")
            print(f"ADJUSTED BUDGET: €{budget:.2f}")
            print(f"(ADJUSTED SHOP FLOAT: €{shop.cash:.2f})")
            print(f"TOTAL BILL SO FAR: €{totalBill:.2f}")

        elif choice == 10:
            print("\n--------------------\n")
            print(f"Your total bill is €{totalBill:.2f}\n")
            print("Thank you for your custom. Please come again soon!\n")
            print("--------------------\n")
            break

        else:
            print("\n** Invalid entry - please try again! **\n")

    shop.cash += totalBill

def mainMenu():
    print("Welcome to the python proc shop\n")
    print("Please select from the following: \n")
    print("1 - Show shop's current stock and float")
    print("2 - John's shopping list")
    print("3 - Tony's shopping list")
    print("4 - Bob's shopping list")
    print("5 - Shop in Live Mode")
    print("0 - Exit\n")

def shopMenu(shop):
    mainMenu()

    while True:
        choice = input("Please make a selection: ")

        if choice == "1":
            print_shop(shop)
            mainMenu()

        elif choice == "2":
            customer1 = read_customer("G00232918_James_Connolly_Assignment_1\customer1.csv")
            print_customer(customer1, shop)
            mainMenu()

        elif choice == "3":
            customer2 = read_customer("G00232918_James_Connolly_Assignment_1\customer2.csv")
            print_customer(customer2, shop)
            mainMenu()

        elif choice == "4":
            customer3 = read_customer("G00232918_James_Connolly_Assignment_1\customer3.csv")
            print_customer(customer3, shop)
            mainMenu()

        elif choice == "5":
            print("You are now entering our LIVE SHOPPING MODE!\n")
            liveMode(shop)

        elif choice == "0":
            print("\n** Bye! Thanks for your custom! **\n** Come again soon! **\n")
            exit()

        else:
            print("** Invalid entry - please enter a number between 0 and 5! **")
            mainMenu()

def main():
    newShop = create_and_stock_shop()
    shopMenu(newShop)

if __name__ == "__main__":
    main()

File: G00232918_James_Connolly_Assignment_1/c/shop.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>
#include <ctype.h>


// Structs to be used for the shop simulation
struct Product {
    char* name;
    double price;
};

struct ProductStock {
    struct Product product;
    int quantity;
};

struct Shop {
    double cash;
    struct ProductStock stock[25];
    int index;
};

struct Customer {
    char* name;
    double budget;
    struct ProductStock shoppingList[30];
    int index;
};

// Create global variables
// (GeeksforGeeks, 2022)
struct Customer c;
struct Shop s;

// Printing product info
void printProduct(struct Product p) {
    printf("-------------\n");
    printf("PRODUCT NAME: %s \nPRODUCT PRICE: %.2f\n", p.name, p.price);
    printf("-------------\n");
}

// Reads in the stock list
struct Shop createAndStockShop() {
    FILE* fp;
    char* line = NULL;
    size_t len = 0;
    size_t read;

    // Opens the stock list from the stock csv
    fp = fopen("C:\\Users\\james\\College\\Multi_Paradigm_Prog\\G00232918_James_Connolly_Assignment_1\\stock.csv", "r");
    
    if (fp == NULL) {
        // (GeeksforGeeks, 2017a)
        perror("Cannot open the stock file");
        exit(EXIT_FAILURE);
    }

    // Read the cash value for the shop
    //read = getline(&line, &len, fp);
    //float cash = atof(line);

    read = getline(&line, &len, fp);
    double shopFloat = atof(line);
    s.cash = shopFloat;


    // Intialise the shop with the cash
    //struct Shop shop = { cash };
    // Parsing each line of the stock file
    while ((read = getline(&line, &len, fp)) != -1) {
        char* n = strtok(line, ",");
        char* p = strtok(NULL, ",");
        char* q = strtok(NULL, ",");
        int quantity = atoi(q);
        double price = atof(p);
        char* name = malloc(sizeof(char) * 50);
        strcpy(name, n);
        struct Product product = { name, price };
        struct ProductStock stockItem = { product, quantity };
        s.stock[s.index++] = stockItem;

    }


    // Close the file and return the shop
    //return shop;
    return s;
}

// List of items in stock with quantities
void printShop() {
    // Print info about the shop's stock
    printf("In stock list\n");
    printf("Shop has %.2f in cash\n", s.cash);
    for (int i = 0; i < s.index; i++)
    {
        // Print the details of each product
        printProduct(s.stock[i].product);
        printf("The shop has %d of the above\n", s.stock[i].quantity);
    }
}

// Check if the product is in stock
int checkProductStock(char* n, int order) {
    for (int i = 0; i < s.index; i++) {
        // Checking
        struct ProductStock product = s.stock[i];
        char* name = product.product.name;
        if (strcmp(name, n) == 0) {
            if (s.stock[i].quantity >= order) {
                // If quantity exists, update the stock and return the order
                s.stock[i].quantity = s.stock[i].quantity - order;
                return order;
            } else if (order > s.stock[i].quantity) {
                // If there is not enough stock provide the update
                order = s.stock[i].quantity;
                printf("---------------\n");
                printf("** We don't have that much %s in stock. We only have %d **\n", product.product.name, s.stock[i].quantity);
                // This is to set the stock to 0 if the customer tries to purchase more than stock
                s.stock[i].quantity = 0;
                }
            return order;
        }
    }
    // Return an error if the stock is not found
    return -1;
}

// finds price of product
double findProductPrice(char *n) {
	for (int i = 0; i < s.index; i++) {
		struct Product product = s.stock[i].product;
		char *name = product.name;
		if (strcmp(name, n) == 0) {
			return product.price;
		}
	}
	return -1;
}

// Reads in customer.csv file
// (Stack Overflow, 2020)
struct Customer createCustomer(char* path_to_customer) {
    FILE* fp;
    char* line = NULL;
    size_t len = 0;
    size_t read;
    c.index = 0;
    // Insert customer csv files
    fp = fopen(path_to_customer, "r");
    if (fp == NULL) {
        printf("Can't access customer file. Please try again!");
        exit(EXIT_FAILURE);
    }
    // Reads the customer's name and budget
    read = getline(&line, &len, fp);
    char* n = strtok(line, ",");
    char* b = strtok(NULL, ",");
    char* custName = malloc(sizeof(char) * 50);
    strcpy(custName, n);
    double custBudget = atof(b);
    c.name = custName;
    c.budget = custBudget;
    struct ProductStock custShopList;

    // Read each line to populate the customer list
    while ((read = getline(&line, &len, fp)) != -1) {
        char* n = strtok(line, ",");
        char* q = strtok(NULL, ",");
        char* name = malloc(sizeof(char) * 50);
        strcpy(name, n);
        int quantity = atoi(q);
        custShopList.product.name = name;
        custShopList.product.price = findProductPrice(name);
        custShopList.quantity = quantity;
        c.shoppingList[c.index++] = custShopList;
    }

    return c;
}

// *** LIVE MODE ***
void liveMode() {
    // Total bill had to be set to 0.0 as the program was going into
    // continous loop when selecting option 5
    double myBudget, totalCost, totalBill = 0.0;
    int quantity_requested, quantity_chosen, select = 0; // Initialize 'select' variable
    printf("What is your budget?\n");
    scanf("%lf", &myBudget);
    do {
        printf("\n***********************\n");
        printf("Welcome to the C shop LIVE SHOPPING MODE!\n");
        printf("***********************\n\n");
        printf("Your current budget is €%.2f.\n\nPlease select what you would like to buy from the list below:\n\n", myBudget);
        for(int i=0; i < s.index; i++) {
            printf("%d - %s @ €%.2f each.\n", i + 1, s.stock[i].product.name, findProductPrice(s.stock[i].product.name));
        }
        printf("Enter your choice (23 to print total bill, 24 to exit): ");
        scanf("%d", &select); // Read the user's selection

        switch(select) {
            case 23: {
                printf("Your total bill is €%.2f.\n", totalBill);
                printf("Thank you for your custom. Please come again soon!\n");
                break;
            }
            case 24: {
                printf("Come again soon and have a nice day!\n");
                return;
                break;
            }
            default: {
                if (select > s.index + 1 || select <= 0) {
                    printf("\n** Invalid entry - please try again! **\n");
                    break;
                }

                printf("How many %s would you like to purchase? ", s.stock[select - 1].product.name);
                scanf("%d", &quantity_requested);
                quantity_chosen = quantity_requested;

                totalCost = quantity_chosen * findProductPrice(s.stock[select - 1].product.name);

                if (myBudget < totalCost) {
                    quantity_chosen = myBudget / findProductPrice(s.stock[select - 1].product.name);
                }

                quantity_chosen = checkProductStock(s.stock[select - 1].product.name, quantity_chosen);

                myBudget -= totalCost;
                totalBill += totalCost;
                s.cash += totalCost;

                printProduct(s.stock[select - 1].product);
                printf("---------------\n");
                printf("QUANTITY REQUIRED: %d\n", quantity_requested);
                printf("QUANTITY PURCHASED: %d\n", quantity_chosen);
                printf("TOTAL ITEM COST: €%.2f\n", totalCost);
                printf("ADJUSTED BUDGET: €%.2f\n", myBudget);
                printf("(ADJUSTED SHOP FLOAT: €%.2f\n", s.cash);
                printf("TOTAL BILL SO FAR: €%.2f\n", totalBill);
                printf("---------------\n");

                if (quantity_requested != quantity_chosen) {
                    printf("\n** You don't have enough money for %s! **\n", s.stock[select - 1].product.name);
                }
            }
        }
    // This is to allow the code to continue executing once these cases
    // are not selected    
    } while (select != 23 && select != 24);
}


// Print info about each customer and their shopping list
// Bool used for the customer shopping, if its false it skips the function
// (Deshmukh, 2021)
void printCustomer(bool upd) {
    printf("%s's shopping list\n", c.name);
    printf("BUDGET: €%.2f\n", c.budget);
    int QC, quantity_chosen;
    double cost, totalBill = 0;
    // Loop through each item in the shopping list
    for (int i = 0; i < c.index; i++) {
        // Get the quantity chosen by the customer
        quantity_chosen = c.shoppingList[i].quantity;
        if (c.shoppingList[i].product.price > -1) {
            // Calculate the cost of product based on the quantity chosen
            cost = quantity_chosen * c.shoppingList[i].product.price;
            if (upd) {
                QC = checkProductStock(c.shoppingList[i].product.name, quantity_chosen);
                // Adjust the quantity chosen based on the available stock
                cost = QC * c.shoppingList[i].product.price;
                if (cost > c.budget) {
                    printf("** Sorry! You don't have enough money left for %s! **\n", c.shoppingList[i].product.name);
                    printf("---------------\n");
                    continue;
                }
                // Update each
                c.budget -= cost;
                s.cash += cost;
                totalBill += cost;
                printf("---------------");
                printProduct(c.shoppingList[i].product);
                printf("QUANTITY REQUIRED: %d\n", c.shoppingList[i].quantity);
                printf("QUANTITY PURCHASED: %d\n", QC);
                printf("TOTAL ITEM COST: €%.2f\n", cost);
                printf("ADJUSTED BUDGET: €%.2f\n", c.budget);
                printf("(ADJUSTED SHOP FLOAT: €%.2f)\n", s.cash);
                printf("TOTAL BILL SO FAR: €%.2f\n", totalBill);
                printf("---------------");
            }
        }
    }
    printf("\nTOTAL BILL: €%.2f\n", totalBill);
    printf("BUDGET REMAINING: €%.2f\n", c.budget);
    printf("\n** Thank you for your custom **\n\n");
}


// Main Menu with the options to select from
//(Stack Overflow, 2015)
void mainMenu(struct Shop s) {
    int menuSelect;
    do {
        printf("\n***********************\n");
        printf("WELCOME TO THE C SHOP\n");
        printf("***********************\n");
        printf("\nPlease select from the following:\n\n");
        printf("1 - Show shop's current stock and float\n");
        printf("2 - John's shopping list\n");
        printf("3 - Tony's shopping list\n");
        printf("4 - Bob's shopping list\n");
        printf("5 - Shop in Live Mode\n");
        printf("0 - Exit\n");
        printf("\nPlease make a selection: ");
        scanf("%d", &menuSelect);
        // Switch allows to select each case and execute it
        switch (menuSelect) {
            case 1: {
                printShop(s);
                break;
            }
            case 2: {
                struct Customer customer1 = createCustomer("C:\\Users\\james\\College\\Multi_Paradigm_Prog\\G00232918_James_Connolly_Assignment_1\\customer1.csv");
                printCustomer(true);
                break;
            }
            case 3: {
                struct Customer customer2 = createCustomer("C:\\Users\\james\\College\\Multi_Paradigm_Prog\\G00232918_James_Connolly_Assignment_1\\customer2.csv");
                printCustomer(true);
                break;
            }
            case 4: {
                struct Customer customer3 = createCustomer("C:\\Users\\james\\College\\Multi_Paradigm_Prog\\G00232918_James_Connolly_Assignment_1\\customer3.csv");
                printCustomer(true);
                break;
            }
            case 5: {
                printf("\n////////////////////////////////////////////");
                printf("\nYou are now entering our LIVE SHOPPING MODE!\n");
                printf("\n////////////////////////////////////////////");
                liveMode();
                break;
            }
            case 0: {
                printf("\n** Bye! Thanks for your custom! **\n** Come again soon! **\n\n");
                break;
            }
            default: {
                printf("** Invalid entry - please enter a number between 0 and 5! **\n");
                break;
            }
        }
    // if the selection is not 0, it is true    
    } while (menuSelect != 0);
}

// main method
int main() 
{
    
    // create shop
    struct Shop newShop = createAndStockShop();
    // newShop is the representation if each option chosen
    mainMenu(newShop);
    return 0;
}


// References
// GeeksforGeeks. (2022). Global Variables in C. [online] Available at: 
// https://www.geeksforgeeks.org/global-variables-in-c [Accessed 15 Nov. 2023].
// GeeksforGeeks. (2017a). Error Handling in C. [online] Available at: 
// https://www.geeksforgeeks.org/error-handling-in-c/ [Accessed 18 Nov. 2023].
// Stack Overflow. (2020). C -Print contents of a csv file. [online] Available at: https://stackoverflow.com
// questions/60937894/c-print-contents-of-a-csv-file [Accessed 16 Nov. 2023].
// Deshmukh, A. (2021). Boolean in C with Examples. [online] Scaler Topics. 
// Available at: https://www.scaler.com/topics/boolean-in-c/ [Accessed 22 Nov. 2022].
// Stack Overflow. (2015). Simple Option Menu (C). [online] Available at: https://stackoverflow.com/questions/
// 33929671/simple-option-menu-c [Accessed 21 Nov. 2023].